Cython/Compiler/ExprNodes.py

```diff
@@ -2055,7 +2064,12 @@ def analyse_target_declaration(self, env):
             self.entry = env.lookup_here(self.name)
         if not self.entry and self.annotation is not None:
             # name : type = ...
-            self.declare_from_annotation(env, as_target=True)
+            if 'dataclasses.dataclass' in env.directives:
+                # in a dataclass an assignment should not prevent a name becoming a class member
+                # hence "as_target = False"
+                self.declare_from_annotation(env, as_target=False)
+            else:
+                self.declare_from_annotation(env, as_target=True)
```

Comment:
Just assign `is_dataclass = 'dataclasses.dataclass' in env.directives` and pass that to `as_target`. No need to duplicate the calls. (The more code I have to read in order to spot the difference, the more I get distracted while trying to understand the code. Don't make me think, just give it a name.)